### PR TITLE
Add StructuredDataInitializable default implementation

### PR DESCRIPTION
**Pull request by bradhilton**
This pull request adds a basic default implementation for `StructuredDataInitializable` using `Reflection`. It also adds `StructuredDataInitializable` implementations for `Optional`, `Bool`, `Double`, `Int`, `String`, `Data`, `Array` and `Dictionary`.

### CODE REVIEW

**Comment by Danappelxx** [comment]
This is awesome! I do think that this should be over at [Zewo/Mapper](https://github.com/Zewo/Mapper) though since it's higher level stuff.

**Comment by bradhilton** [comment]
In my discussion with @paulofaria, we decided to provide this as a default implementation. I think it makes the most sense to have the default implementation for `StructuredDataInitializable` live in `StructuredData`.

**Comment by bradhilton** [comment]
I also proposed as an alternative a new type, that inherits from `StructuredDataInitializable`, something like `StructuredDataModelInitializable`, but we opted for the simpler route.

File: Sources/Reflection.swift [hunk 1]
```diff
@@ -0,0 +1,122 @@
+import Reflection
+
+extension C7.StructuredDataInitializable {
+    
+    init(structuredData: StructuredData) throws {
```
Comment (Danappelxx) [inline Sources/Reflection.swift]: should be public

**Comment by Danappelxx** [comment]
That makes sense but I'm reluctant to depend on reflection in such a core module. This seems higher level to me (hence Mapper), and although Reflection is totally awesome, it's also totally unstable.

File: Sources/Reflection.swift (hunk 1 above)
Comment (bradhilton) [inline Sources/Reflection.swift]: Good catch

**Comment by bradhilton** [comment]
I'm fine with moving it somewhere else. Maybe `Mapper` is the right place. I'd like to have @paulofaria 's input.

I'd also like to point out that the ABI is unstable for everything, every new version of Swift has breaking changes, so until Swift releases are actually stable, it's a moot point.

**Comment by bradhilton** [comment]
Look like it's failing due to a swift env issue, because C7 is on 06-06 and we're on 05-09. Any idea why this wasn't failing before? We'll have to get in sync with C7 before this will work.

**Comment by paulofaria** [comment]
I agree with @bradhilton that stability is not something we're living with these days anyway. Having `Reflection` as an active part of `StructuredData` makes more sense in the way that it's just a default implementation for `StrucutedDataInitializable` and because we obviously can't put it in `C7`, `StructredData` seems to be the right place. And actually about `Mapper`, I think we can merge `Validator` and `Mapper` and make `Mapper` really more than just boilerplate. Try to make it be used with really more complex mappings which need validation, etc.

**Comment by paulofaria** [comment]
And of course if `Reflection` breaks, @bradhilton will be on call 24h in our support line, hahaha.

**Comment by bradhilton** [comment]
Hahaha of course 😄

**Pull request closed without merging**